### PR TITLE
RE2022-336: replace threads and program_threads in taskfarmer task generator

### PR DESCRIPTION
**Pull request by Tianhao-Gu**
With the changes, current tool container will break until we update them (which should happen soon after this PR).

### CODE REVIEW

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 1]
```diff
@@ -36,29 +36,32 @@
...
+# chunk_size is the quantity of genomes processed within each task (default is 5000)
+#    for batch genome tools, such as gtdb_tk and checkm2, the chunk_size specifies the number of genomes grouped
+#    together for processing in a batch by the tool
+#    for single genome tools, such as microtrait and mash, the chunk_size is the number of genomes to process in a
+#    serial manner
+# exe_time is the estimated execution time for a single task (default is 60 minutes)
+# tasks_per_node is the number of parallel tasks to run on a node (default is 1)
+# node_time_limit is the time limit for the node we reserved for the task (default is 5 hours)
+# if no specific metadata is provided for a tool, the default values are used.
 TASK_META = {'gtdb_tk': {'chunk_size': 1000, 'exe_time': 65, 'tasks_per_node': 4},
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:61]: Will slurm keep running jobs on a node until it runs out of time?

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 2]
```diff
@@ -36,29 +36,32 @@
...
              'eggnog': {'chunk_size': 100, 'exe_time': 15, 'node_time_limit': 0.5},  # Memory intensive tool - reserve more nodes with less node reservation time
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:62]: Should threads per tool be specified here as well?

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 3]
```diff
@@ -36,29 +36,32 @@
...
+# Task metadata - tool specific parameters for task generation and execution
+# chunk_size is the quantity of genomes processed within each task (default is 5000)
+#    for batch genome tools, such as gtdb_tk and checkm2, the chunk_size specifies the number of genomes grouped
+#    together for processing in a batch by the tool
+#    for single genome tools, such as microtrait and mash, the chunk_size is the number of genomes to process in a
+#    serial manner
+# exe_time is the estimated execution time for a single task (default is 60 minutes)
+# tasks_per_node is the number of parallel tasks to run on a node (default is 1)
+# node_time_limit is the time limit for the node we reserved for the task (default is 5 hours)
+# if no specific metadata is provided for a tool, the default values are used.
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:60]: I assume next is redoing the calcs for the number of nodes and asking for user confirmation?

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 1 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py:61]: I think once jobs are all finished, the node will terminate.

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 2 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py:62]: It's added as a CLI input. and defaults to 32. We should do it here instead?

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py:60]: yea. I want to fix container first before that.

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 2 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py:62]: We're specifying everything else here, why not that?

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 1 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py:61]: :+1:

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py:60]: :+1:

Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py:60]: wait, `number of nodes` calculation is still correct. We don't need to redo that.

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 2 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py:62]: 👍

Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py:62]: No custom per tool settings needed ?

Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py:62]: no. leaving 32 as default for every tool works fine. That's the config we've been using.

Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py:62]: :+1:

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 4]
```diff
@@ -37,28 +37,32 @@
...
-or 16,000 genomes, per node within the 5-hour time limit. We plan to make these parameters configurable based on 
-the specific tool being used. After conducting performance tests, we found that utilizing 32 cores per batch and 
-running 4 batches in parallel per NERSC node resulted in optimal performance, despite each node having a total of 
-256 cores.
 '''
 
 TOOLS_AVAILABLE = ['gtdb_tk', 'checkm2', 'microtrait', 'mash', 'eggnog']
 
-# estimated execution time (in minutes) for each tool to process a chunk of data
+NODE_TIME_LIMIT_DEFAULT = 5  # hours  # TODO: automatically calculate this based on tool execution time and NODE_THREADS
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py]: How can you calculate this and num nodes? If you have to calculate both you're under constrained I think

Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py]: yea. You are correct. The `todo` was blindly copied. It's no longer valid.

Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: :+1:

**Review by MrCreosote — APPROVED** [review]

**Pull request merged**